File: backend/app/api/v1/routes/leaderboard.py
"""Leaderboard route."""

from fastapi import APIRouter

from app.api.v1.deps import CurrentUser, DbSession
from app.schemas.leaderboard import LeaderboardResponse
from app.services import leaderboard_service

router = APIRouter(prefix="/leaderboard", tags=["leaderboard"])


@router.get(
    "",
    response_model=LeaderboardResponse,
    summary="Learner standings by lifetime XP",
    responses={503: {"description": "The database has not been seeded"}},
)
def get_leaderboard(db: DbSession, user: CurrentUser) -> LeaderboardResponse:
    """Return learners ranked by lifetime XP, with the caller's row flagged.

    The caller is resolved from the current-user dependency, never from a query
    parameter, so a client cannot ask to be highlighted as somebody else.
    """
    return leaderboard_service.standings(db, user)

File: frontend/lib/api/leaderboard.ts
/**
 * Leaderboard endpoint.
 *
 * No user id in the signature: the server resolves the caller and flags their
 * row, so a client cannot ask to be highlighted as somebody else.
 */

import { request } from "./client.ts";
import type { LeaderboardResponse } from "./types";

/** Learner standings, ranked by lifetime XP. */
export function getLeaderboard(init?: RequestInit): Promise<LeaderboardResponse> {
  return request<LeaderboardResponse>("/leaderboard", init);
}
